Automattic/woocommerce-payments
# Disable Stripe Billing Subscriptions while in Subscriptions staging mode

Fixes #6316

#### Changes proposed in this Pull Request

When a site is in staging mode, we don't want subscription changes on that staging (dev) site to communicate with the WC Pay server and potentially impact live account data.

Prior to this PR, it was possible that a duplicate site could send subscription-related requests to the WC Pay server, for those requests to be actioned, and then for the live store to be impacted by those changes. 

The most common outcome was as follows:

1. Subscriptions on the staging site are treated as manual due to the duplicate site logic which turns all subscriptions on staging sites into manual subscriptions. 
2. When the scheduled actions for those subscriptions would run, the subscription would go on-hold. 
3. The WC Pay Subscriptions code would hook into this status transition and communicate with Stripe that this subscription was put on-hold. 
4. Stripe would pause the subscription and all future renewals would essentially halt. 

This PR addresses this by:

1. Making sure that we don't hook onto any subscription-status change actions on staging sites. This will prevent subscriptions which change status on duplicate sites from sending corresponding requests to server, essentially acting on behalf of the live site. 
2. All subscription-product related actions have also been disabled. Prior to this, if an admin changed the product name or description, or deleted the product, this would impact the products at Stripe which are used by the live site.  
3. All subscription-related webhooks are also disabled. If a staging site somehow receives a subscription `invoice.upcoming`, `invoice.paid`, or `invoice.failed` webhook, it won't process it. Staging sites should not receive webhooks (they should be forwarded to the live site URL), however this is necessary because our webhook handling can lead to responses. ie we can pause subscriptions at Stripe if we process a `invoice.upcoming` and if the subscription in WooCommerce is cancelled. 
4. Crucially, I've updated the `is_wcpay_subscription()` function to ensure that it returns `false` on staging sites. A lot of our WCPay Subscription logic is built on this check. If we make sure that wcpay subscriptions aren't treated as wcpay subscriptions on staging sites, they should behave like tokenised wc pay subscriptions and avoid a lot of Stripe billing interactions. 
5. Updating `WC_Payments_Invoice_Service` so it won't hook onto any callback and fire off server requests based on order status changes. If an unpaid subscription-related order was marked as processing or complete on a staging site, that would have fired off a request to server to mark the invoice as paid, and activate the subscription. This shouldn't occur on staging sites. 

- I've left the `WC_Payments_Subscription_Change_Payment_Method_Handler` unchanged bar a minor change. This class doesn't do anything except change urls, button text and display a notice when the user clicks the change payment method button. As long as we don't act on the actual changing of the payment method, we should be fine. This will need to tested. 

#### Testing instructions

- Testing the [wcpay subscription critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) should be done. 
- Every WCPay Subscription flow should be tested as if the site is live mode.
- Every flow should then be tested in staging mode. 
- The flows to test are: 
   1. **Products**
      - Create a subscription product.
      - Delete a product
      - Update a product's name.  
   3. **Subscriptions**
       - Purchase a subscription.
       - Suspend a subscription
       - Activate a subscription. 
       - Cancel a subscription.
       - Change a subscription payment method. 
   4. **Renewals**
       - Process successful renewals. 
       - Fail renewals
       - Pay for failed renewals.
       - Run subscription scheduled actions. 
           - On staging this will create pending renewal orders, as long as this doesn't impact the subscription in the Stripe dashboard, this is expected. 
           - While in live mode, running the scheduled action will do nothing. 

**Between each test check that no changes have propagated to the Stripe account while in staging mode. Note that some flows might even be disabled on the staging environment.**

-------------------


**Post merge**
